statistics/handle/lru_cache.go

```diff
@@ -0,0 +1,336 @@
+}
+
+// internalLRUCache is a simple least recently used cache
+type internalLRUCache struct {
+	sync.RWMutex
+	capacity int64
+	// trackingCost records the tracking memory usage of the elements stored in the internalLRUCache
+	// trackingCost should be kept under capacity by evict policy
+	trackingCost int64
+	// totalCost records the total memory usage of the elements stored in the internalLRUCache
+	totalCost int64
+	elements  map[int64]*list.Element
+	// cache maintains elements in list.
+	// Note that if the element's trackingMemUsage is 0, it will be removed from cache in order to keep cache not too long
+	cache *list.List
+}
+
+// newInternalLRUCache returns internalLRUCache
+func newInternalLRUCache(capacity int64) *internalLRUCache {
+	if capacity < 1 {
+		panic("capacity of LRU Cache should be at least 1.")
+	}
+	return &internalLRUCache{
+		capacity: capacity,
+		elements: make(map[int64]*list.Element),
+		cache:    list.New(),
+	}
+}
+
+// Get tries to find the corresponding value according to the given key.
+func (l *internalLRUCache) Get(key int64) (*statistics.Table, bool) {
+	l.Lock()
+	r, hit := l.get(key)
+	l.Unlock()
+	if hit {
+		metrics.StatsCacheLRUCounter.WithLabelValues(typHit).Inc()
+	} else {
+		metrics.StatsCacheLRUCounter.WithLabelValues(typMiss).Inc()
+	}
+	return r, hit
+}
+
+func (l *internalLRUCache) get(key int64) (*statistics.Table, bool) {
+	element, exists := l.elements[key]
+	if !exists {
+		return nil, false
+	}
+	l.cache.MoveToFront(element)
+	return element.Value.(*cacheItem).value, true
+}
+
+// Put puts the (key, value) pair into the LRU Cache.
+func (l *internalLRUCache) Put(key int64, value *statistics.Table) {
+	l.Lock()
+	l.put(key, value, value.MemoryUsage(), true)
+	trackingCost := l.trackingCost
+	totalCost := l.totalCost
+	l.Unlock()
+	metrics.StatsCacheLRUCounter.WithLabelValues(typUpdate).Inc()
+	metrics.StatsCacheLRUMemUsage.WithLabelValues(typTrack).Set(float64(trackingCost))
+	metrics.StatsCacheLRUMemUsage.WithLabelValues(typTotal).Set(float64(totalCost))
+}
+
+func (l *internalLRUCache) put(key int64, value *statistics.Table, tblMemUsage *statistics.TableMemoryUsage, tryEvict bool) {
+	// If the item TotalColTrackingMemUsage is larger than capacity, we will drop some structures in order to put it in cache
+	for l.capacity < tblMemUsage.TotalColTrackingMemUsage() {
+		for _, col := range value.Columns {
+			col.DropEvicted()
+			tblMemUsage = value.MemoryUsage()
+			if l.capacity >= tblMemUsage.TotalColTrackingMemUsage() {
+				break
+			}
+		}
+	}
+	defer func() {
+		if tryEvict {
+			l.evictIfNeeded()
+		}
+	}()
+	element, exists := l.elements[key]
+	if exists {
+		oldMemUsage := element.Value.(*cacheItem).tblMemUsage
+		element.Value.(*cacheItem).value = value
+		element.Value.(*cacheItem).tblMemUsage = tblMemUsage
+		l.calculateCost(tblMemUsage, oldMemUsage)
+		l.maintainList(element, nil, tblMemUsage.TotalColTrackingMemUsage(), oldMemUsage.TotalColTrackingMemUsage())
```

Comment:
Is it possible that `oldMemUsage.TotalColTrackingMemUsage() == 0` and `newMemUsage.TotalColTrackingMemUsage() > 0`? If it is possible, we will add `nil` into list.